internal/directory_contents.go

```diff
@@ -18,36 +18,50 @@ package internal
 
 import (
 	"fmt"
+	"io"
 	"os"
 	"path/filepath"
-	"sort"
+
+	"github.com/buildpacks/libcnb/log"
 )
 
-// DirectoryContents is used to generate a collection of the names of all files within a directory.
-type DirectoryContents struct {
-	Path string
+// DirectoryContentsWriter is used write the contents of a directory to the given io.Writer
+type DirectoryContentsWriter struct {
+	path   string
+	format log.DirectoryContentFormatter
+	writer io.Writer
 }
 
-// Get returns the names of all files within a directory
-func (d DirectoryContents) Get() ([]string, error) {
-	var contents []string
+// NewDirectoryContentsWriter returns a new DirectoryContentsWriter initialized and ready to be used
+func NewDirectoryContentsWriter(path string, format log.DirectoryContentFormatter, writer io.Writer) DirectoryContentsWriter {
```

Comment:
I'm curious what your thoughts are about New* functions. Typically they make it really hard to extend the API if we add new fields to the struct.

Reply:
On one hand, it's an established pattern in Go so I try to follow it.

I do agree that it creates challenges though. It gets unwieldy if you have more than a few arguments, and you're right it's tied very closely to the struct properties so if you change one you have to change the other. It definitely makes it harder to evolve the code without breaking the API.

What I've done in some of our Paketo code is:
    - if we add new fields, we end up adding a new New* method, and deprecating the old method. It's still there, but not breaking.
    - In some cases, we use the Option pattern, where you pass in a variable number of Options that then get applied against the struct

The Option pattern is nice with large structs that have lots of properties or to add more flexibility. The downside is that enforcing required properties then becomes a runtime check as opposed to a requirement through the function arguments.

In this particular case, I added the New method because I wanted to make the struct properties all private (I know you've mentioned wanting to do this more). This would seem to necessitate a constructor, so the struct can be created outside of its package. I'm also OK in this case as I don't think the struct properties will really need to change here. We're already passing in pretty much all the info we can for use in by the formatter.